Allow manifest processing to be invoked without going through an HTML document

mgiuca opened the pull request:
Closes #1068

This change (choose at least one, delete ones that don't apply):

* Adds new normative recommendations or optional items
* Makes editorial changes (changes informative sections, or changes normative sections without changing behavior)

Commit message:

Allow manifest processing to be invoked without going through an HTML document.

Manifest processing: Replaces the link and response parameters with document URL and manifest URL. This was limiting the ability to call the processing algorithm from outside an HTML document context.

Note that the only call to this algorithm is in the HTML spec, which needs to be updated simultaneously to use the new interface.
 
Adds new normative text allowing user agents to invoke the processing steps without a document, provided that
they supply a valid document URL and if the document URL is on a different origin to the manifest, there is a page on the document's origin that links to the manifest.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/mgiuca/manifest/pull/1069.html" title="Last updated on Apr 27, 2023, 3:04 AM UTC (3d03340)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/manifest/1069/0252939...mgiuca:3d03340.html" title="Last updated on Apr 27, 2023, 3:04 AM UTC (3d03340)">Diff</a>

mgiuca commented:
@marcoscaceres @dmurph adding both; let's take our time with this one.

I'm not sure if you want to just start with the interface change and add the text later, if the text is controversial. I wanted to capture that this is explicitly allowed now (invoking this from outside a document) but I also want to make sure user agents verify the bidirectional link. This is not just a security consideration, it is a requirement. (It could also be downgraded to a security consideration if you like.)

It was difficult to phrase the requirement. Note that I don't expect user agents to literally verify the document links to the manifest at install time (otherwise we effectively require that you load the whole document), just that it is known to have linked to the manifest in the past. This means, e.g. if you're doing sync, then you install through the normal HTML flow on device X, recording both the document URL and manifest URL in the sync state, then when you sync to device Y, the user agent on Y can invoke the processing steps with the stored document URL and manifest URL, knowing that they had been bidirectionally associated in the past. Do you think the text adequately captures this?

mgiuca commented:
Wrote whatwg/html#8754 to update HTML to use the new interface. (Which also moves some of the deleted lines from Manifest into HTML.)

marcoscaceres reviewed (approved):
I'm still a bit worried about the security implications of this (in that the document provides the security/cors policy), so I wonder if we should add a note that this is a bit experimental?

dmurph reviewed (approved):
This LGTM - the other patch is approved - can we modify the descriptions & submit?

aarongustafson commented:
LGTM as well. I welcome this change! It'll be a boon for app stores, packaging, & so much more. Thanks for doing this @mgiuca!

mgiuca commented:
Hi folks,

Sorry for taking awhile to circle back on this. I wanted to take time to self-review since it was a long time since I uploaded the PRs, and that meant pushing it back until I found some free time.

So now that I've paged it back in, the security concern Marcos had is basically that the HTML applies lots of CORS checks before downloading the manifest. This change (particularly, what's implied by the non-normative note) allows user agents to magically supply a manifest from anywhere, not follow the HTML download steps. In some sense, that's the whole point (e.g., we are trying to use this to create a service that installs apps from pre-cached manifests, we explicitly _do not want_ to have to follow the HTML download steps to fetch it from its actual URL). But in another sense, that opens up the possibility of a UA not following CORS.

It's possible we want stronger language, like "the user agent MUST use a manifest that it previously downloaded using the CORS stuff from HTML" but I don't really know how to phrase that or what the implications are. For now, I think (if you all agree) it's OK to leave that up to user agents to work out, and trust that they will avoid spoofs. Maybe this can just be a security consideration (in a follow-up? Or should I add it in this PR?).

Since several of you have pinged me to say "please merge this so we can get on with our lives", I'm thinking we just merge it for now and add security considerations as a followup. Unless someone objects, I'll do that as soon as I get an approval on whatwg/html#8754. I have rebased.

****

I haven't used "merge queue" before. I don't see a way that I can enter a commit message -- how do you do that now? (Is it going to use the original post text? That contains stuff I don't want including the preview and draft links.)

aarongustafson commented:
@mgiuca Perhaps adding another RECOMMENDED to the [Privacy & Security Considerations](https://www.w3.org/TR/appmanifest/#priv-sec) section that recommends UAs that make use of Manifest files not bound to an HTML document apply the same rigor & protections they would under CORS? Since it's possible the Manifest could be part of a package, I'm not really sure what the mechanism should be. How does the Web Packaging spec deal with this?

As for the merge queue, I have no idea. I don't think I've paid attention to that before, I've mainly squashed (which I thought was what we had on this repo previously).

mgiuca commented:
Thanks @aarongustafson - I'm re-reading up on CORS now so I can do it justice.

I filed #1075 about the merge queue thing.

mgiuca commented:
I did a lot of thinking about CORS and here's what I came up with. Note that I've moved all the new text to its own section, "Processing a manifest without a document" which you can see in the preview. It now basically has prose that says "you SHOULD do CORS the way HTML does it".

Note that this is blocked on whatwg/html#9204 to export the definition of "CORS settings attribute credentials mode".

Also all the links to the link rel "`manifest`" value (both here, and existing already in the document) link to the wrong place due to w3c/respec#4435. Were it not for that, I'd have used `[^link/rel/manifest^]` more to directly link to the algorithm in HTML, but I can't get a direct link to that to work at the moment.

mgiuca commented:
I've done another pass over this section with @alancutter (thanks).

- Changed MUST into a SHOULD. We can't really expect all uses to directly verify this.
- Removed the "or" clause that the document be same-origin as manifest; you still want a link from the document to the manifest either way.
- Added "at least at some point in the past", to acknowledge that you don't need to verify this at install time, just whenever you did the caching.
- Clarify that the CORS request is only needed if the manifest is not same-origin as the document.

I think this basically captures the way that I at least expect this interface to be used outside of HTML. In short, it says "this should be called as-if you found it in an HTML document and invoked the 'steps to process this type of linked resource'". It captures the need to have a link from somewhere on this origin, and make the request with CORS.

Would appreciate another review from at least one of { @marcoscaceres , @dmurph , @aarongustafson } since I've rewritten the prose section in its entirety.

On the plus side, when this was originally proposed we were thinking we'd need a follow up to think about CORS better - that has now been done!

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+            The [=processing a manifest=] steps are invoked by [[HTML]]'s
+            processing steps for the [^link^] element, but MAY also be invoked
+            by the user agent to process a manifest without an associated
+            document.
```
```suggestion
            [=document=].
```

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+            document.
+          </p>
+          <p>
+            In this case, the user agent SHOULD ensure that, at least at some
+            point in the past:
```
```suggestion
            In this case, the user agent SHOULD ensure that at least at some
            point in the past:
```

mgiuca replied:
By the way, the "at least some point in the past" clause was to acknowledge that we don't need the user agent to verify this exactly at install time. It's sufficient to have verified this at the time you cached it (e.g. if you install on Machine A, you verify this, then sync the metadata to Machine B and install it, you don't need to verify it again on Machine B).

We thought about writing "at the time of fetching the metadata" but we wanted to also allow a situation where you verify it on Machine A, then you sync the manifest URL to Machine B, Machine B fetches the manifest but does not re-verify this condition. So we settled for the more permissive "at some point in the past".

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+            point in the past:
+          </p>
+          <ul>
+            <li>there was at least one page which is [=same origin=] as
```
```suggestion
            <li>there was at least one [=document=] which is [=same origin=] as
```

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+          </p>
+          <ul>
+            <li>there was at least one page which is [=same origin=] as
+            |document URL:URL| that has a [^link^] element
+            |linkElement:HTMLLinkElement| with a [^link/rel^] of
+            [^link/rel/manifest^] and a [^link/href^] that resolves to
+            |manifest URL:URL|, and that
```
Nit: not sure if we have a concept of this... but it feels like we should say something like "a valid link type manifest"... or just a "link type "manifest"", as per HTML.

mgiuca replied:
HTML doesn't export a concept of "link type". Internally, they just link "link type" to the [#linkTypes](https://html.spec.whatwg.org/multipage/links.html#linkTypes) heading, but it isn't exported.

I think it's reasonable to keep this the way it is; it's just literally stating what it is instead of referring to a concept.

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+            |manifest URL:URL|, and that
+            </li>
+            <li>if |manifest URL| is not [=same origin=] as |document URL|, the
+            |bodyBytes:byte sequence| data was fetched with a [=CORS Request=]
```
Maybe "is fetched"?

mgiuca replied:
Done.

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+            </li>
+          </ul>
+          <p>
+            (The [[HTML]] steps automatically ensure this.)
```
I wonder if you can can say this directly in the above (something like, "In keeping with [[HTML]] behavior..." or something). It's a bit risky in case it gets moved or whatever, but should be ok... it will also make it more clear what "it" is.

mgiuca replied:
Done.

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+          <p>
+            (The [[HTML]] steps automatically ensure this.)
+          </p>
+          <div class="note">
```
Might be good to add a "`title="some title for the note"`"

mgiuca replied:
Done.

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+          <p>
+            (The [[HTML]] steps automatically ensure this.)
+          </p>
+          <div class="note">
```
```suggestion
          <aside class="note">
```

marcoscaceres commented on index.html:
```diff
@@ -1382,6 +1375,62 @@ <h2>
+              the manifest is fetched with or without credentials, as agreed by
+              both origins, as it would if going directly through the document.
+            </p>
+          </div>
```
```suggestion
          </aside>
```

marcoscaceres reviewed (changes requested):
couple of little things, but otherwise looks good I think.

mgiuca commented:
Done the requested changes (see comments).

mgiuca commented:
Friendly ping, @marcoscaceres - do the updates look good now?

marcoscaceres commented:
Oops! sorry for the delay @mgiuca. Was traveling last week.
